### ISSUE TITLE
[BUG]: Filtering articles into multiple labels causes huge slowdown

### ISSUE THREAD

**Issue by Mitecon** [martinrotter/rssguard#952]
### Brief description of the issue

I make heavy use of filters and labels. Quite often I'm not able to go through everything for a while and so the articles can sit there unattended for some time until I'm able to look at them. Over time, the amount of articles build up and because many of them can be filtered into more than one label, an enormous amount of lag is introduced when doing simple things like opening the filter dialogue, switching labels or even sorting columns alphabetically or by date.

### How to reproduce the bug?

I've been doing some testing and I have created a test database which has only one feed and, for simplicity's sake, I created filters to apply the relevant labels to every article containing a vowel:

[vowels-database.txt](https://github.com/martinrotter/rssguard/files/11432347/vowels-database.txt)

(I can't upload anything with a .db extension so this needs a rename.)

Load up the database, right-click on the feed and 'Fetch selected'. It would be useful to have your task manager open to see how threads become fully locked up during this process. If you have your task manager open to the right you will be able to see the articles arriving into the labels on the left. This will take some time as each article is filtered.

Once the feed has updated and the articles have the labels applied, try clicking through the UI. For instance, open the filter dialogue and time how long it takes. Then try clicking on each filter one by one and see how long that takes. Close this dialogue and try clicking on each label and time how long that takes.

Depending on your system, it may actually not be too slow for you. So I created another test database which has a more extreme set of filters that pick up a few extra common words:

[extreme-database.txt](https://github.com/martinrotter/rssguard/files/11432368/extreme-database.txt)

Obviously, this is a stupid set of filters designed for testing and reproducing this problem that nobody would use day-to-day. However...

The feed I've found only has around 1800 articles in it. Now, imagine dozens of feeds where some of them have articles added regularly throughout each day. Quite quickly this can (and on mine, does) build up to thousands of articles. Now imagine there are filters that sort many of these articles into multiple labels and you can see how this lag will become exponentially longer.

In fact, if you do any of the above with either test database and go into each label, select all articles and remove the labels from them, RSS Guard becomes a little bit faster each time. If you remove all the labels from all articles this way, then navigating the UI become almost instantaneous.

Go into the filter dialogue and apply all the filters again and RSS Guard slows down once again.

So the problem comes from having multiple articles having multiple labels applied. This, for me, is how I work. I do hate having to mention this but my workflow was like this for years with QuiteRSS which has never exhibited this issue even with a many years-old database.

Looking through the RSS Guard database, I can see that articles have separate entries for each label that is applied - meaning the articles are duplicated, sometimes multiple times depending on how many labels are applied, in the database:

![rssguard-db-labels](https://github.com/martinrotter/rssguard/assets/125831943/8d702662-4f8e-4832-8093-fcffb2d76d03)

In QuiteRSS, the articles have the corresponding label ID applied to each article in a single column - meaning each article is a single unique entry in the database with no duplication:

![quiterss-db-labels](https://github.com/martinrotter/rssguard/assets/125831943/21e3084b-bb0f-4443-b11c-3f460913544d)

I obviously understand that to change RSS Guard's database would entail a rather significant rewrite and probably require everyone to wipe out their databases since I imagine a new format would be incompatible.

I also understand that, from what I've read in the Issues, no-one else has reported this problem. So this might just be me who has encountered this and everyone else is sailing by without really making heavy use of labels. But then again, there might be a whole bunch of people who tried RSS Guard, hit this problem, decided it was 'too slow' and went back to their old RSS reader without ever reporting the issue here and so no-one will ever know.

### What was the expected result?

RSS Guard should not have massive lag during simple operations such as navigating the UI.

### What actually happened?

For context, with my current database, it takes more than 3 minutes to open the filter dialogue. The same again to switch to a different label. Or anything, really. This labels problem makes RSS Guard unwieldy for me and I would like for it to be... wieldy.

I realise if I am the only one with this problem then it won't be a high priority. But it's a bug report, so... I wish I was a programmer so I could do something about this. Unfortunately I'm not.

### Debug log

A log is not really useful here. It shows nothing out of the ordinary. Besides, using the databases I've attached above will allow you to test yourself on your own system(s).

### Operating system and version

System 1:
Operating System: Manjaro Linux
KDE Plasma Version: 5.27.4
KDE Frameworks Version: 5.105.0
Qt Version: 5.15.9
Kernel Version: 6.1.26-1-MANJARO (64-bit)
Graphics Platform: X11
Processors: 8 × AMD Ryzen 7 3700X 8-Core Processor
Memory: 15.6 GiB of RAM
Graphics Processor: NVIDIA GeForce GT 1030/PCIe/SSE2
Manufacturer: QEMU
Product Name: Standard PC (Q35 + ICH9, 2009)
System Version: pc-q35-4.2

RSS Guard: v4.3.4 Flatpak

System 2:
Operating System: Manjaro Linux
KDE Plasma Version: 5.27.4
KDE Frameworks Version: 5.105.0
Qt Version: 5.15.9
Kernel Version: 6.1.26-1-MANJARO (64-bit)
Graphics Platform: X11
Processors: 8 × Intel® Core™ i7 CPU Q 720 @ 1.60GHz
Memory: 11.6 GiB of RAM
Graphics Processor: AMD REDWOOD

RSS Guard: v4.3.4 Appimage

System 3:
Device name	Laptop
Processor	Intel(R) Core(TM) i7 CPU       Q 720  @ 1.60GHz   1.60 GHz
Installed RAM	12.0 GB
System type	64-bit operating system, x64-based processor
Pen and touch	No pen or touch input is available for this display
Edition	Windows 10 Home
Version	22H2
Installed on	‎16/‎04/‎2023
OS build	19045.2846
Experience	Windows Feature Experience Pack 120.2212.4190.0

RSS Guard: v4.3.4 ( rssguard-4.3.4-fb2c439b-win10.exe)

System 2 and 3 are the same laptop with two separate hard drives, dual booting.

**Comment by martinrotter** [comment]
Hi.

Thank you for very elaborated report.

First, articles ARE NOT duplicated in RSS Guard database. There is single table "Messages" which holds all articles and each article is SINGLE row in the table.

Label asssigned to articles are stored in separate 1:N table.

I will test your database. There might be some performance killer bug I am not aware of.

**Comment by rfkat** [comment]
Not sure this is related to labels. I've also got an incredible lag when opening the filter dialog (currently 2 minutes 20 seconds for first opening, roughly 45 seconds just for clicking on a different filter in the list and waiting for it to become displayed). My impression is rather that filling the "Existing articles" tab with tens of thousands of articles may not be a good idea.

**Comment by rfkat** [comment]
> You don't say whether you do or don't use labels but if you do, there is an easy test to see if this is the same cause of your slowdown:

Ouch. You are right. Deleting the labels gets me back to normal speed, without any lag.

**Comment by rfkat** [comment]
> What is your operating system, specs and RSS Guard version/type?

Linux, RSS Guard 4.3.4. Not sure what you mean by specs, but an Intel Core i5-10210U with 32GB of RAM available should be sufficient, at least `htop` says I'm nowhere getting near any kind of stress working with RSS Guard.

**Comment by rfkat** [comment]
> Specs just means processor, RAM, etc. I would say yours is more than good enough.

- SSD
- Manjaro Linux (kernel 6.1.26)
- KDE
- RSS Guard full version
- flatpak

**Comment by rfkat** [comment]
> Out of curiosity, can you open System Monitor and screenshot the Process Table tab (put 'rssguard' in the Quick Search box). Make sure you have the Memory and Relative Start Time columns visible in the screenshot.

[https://pasteboard.co/NVjTQplVwqIq.jpg](https://pasteboard.co/NVjTQplVwqIq.jpg)

**Comment by rfkat** [comment]
> Out of curiosity, can you open System Monitor (EDIT: might be KSysGuard) and screenshot the Process Table tab (put 'rssguard' in the Quick Search box). Make sure you have the Memory and Relative Start Time columns visible in the screenshot.

[https://pasteboard.co/HX2d6ySRgkYz.jpg](https://pasteboard.co/HX2d6ySRgkYz.jpg)

**Comment by rfkat** [comment]
> So your instance of RSS Guard has been running (hmm.. timezones... are you in Germany?) for less than two hours and it's using 1.3GB RAM? That's quite a lot for that amount of time. Also, you might not have noticed this before with 32GB RAM but cache pressure would have been an issue for you on, say, 8GB RAM.

At the time I took the screenshot, it had been running for about 6 hours (now almost 8 hours). Mem value as of now keeps jumping between ~930MB and 1,3GB. It's running on my workstation where I keep doing my daily work in parallel, did not notice issues so far.

**Comment by martinrotter** [comment]
OK guys.

I did some preliminary testing and debugging. All problems we see stem from DB layer and its design. SQL layer is designed in a way to be very simple for programmer and follows all basic DB design principles. Are associations (1:M mostly) are made via separate tables, no data are duplicated. Design itself is very clean.

Sadly when data is separated to multiple tables then some queries turn out to tak HUGE amount of time when there is big number of articles AND big number of labels assigned to those articles. This fact did not arose when I was designing the DB some years ago simply because I did not test it with big number of labels.

Here is sample SQL query which simply lists all articles:

```sql
ANALYZE;
SELECT Messages.id,
        Messages.id + 0 as sss,
       Messages.is_read,
       Messages.is_important,
       Messages.is_deleted,
       Messages.is_pdeleted,
       Messages.feed,
       Messages.title,
       Messages.url,
       Messages.author,
       Messages.date_created,
       Messages.contents,
       Messages.enclosures,
       Messages.score,
       Messages.account_id,
       Messages.custom_id,
       Messages.custom_hash,
       Feeds.title,
       CASE WHEN length(Messages.enclosures) > 10 THEN 'true' ELSE 'false' END AS has_enclosures,
       (
           SELECT GROUP_CONCAT(Labels.name) 
             FROM Labels
            WHERE Labels.custom_id IN (
                      SELECT LabelsInMessages.label
                        FROM LabelsInMessages
                       WHERE LabelsInMessages.account_id = Messages.account_id AND 
                             LabelsInMessages.message = Messages.custom_id
                  )
       )
       AS msg_labels
  FROM Messages
       LEFT JOIN
       Feeds ON Messages.feed = Feeds.custom_id AND 
                Messages.account_id = Feeds.account_id
 WHERE Messages.is_deleted = 0 AND 
       Messages.is_pdeleted = 0 AND 
       Messages.account_id = 1
 ORDER BY Messages.is_read ASC;
```

The problem is column `msg_labels` which contains some string operations with `GROUP_CONCANT` operator and this is exectuted for each single row and takes huuuge amount of time. This specific column is responsible for showing titles of labels assigned to article in article list.

There are several similar performance problems caused essentially by same SQL things, all connected to labels. This in effect makes all some operations regarding labels ultra slow.

I will have to fix and check it all and hope that it is all fixable with non-backwards-compatible SQL fixes. If those fixes will be bigger, then new major RSS Guard release will have to be released.

QuiteRSS includes IDs of assigned labels directly in "article" table which is from design POV big "no", but I have to say that performance-wise the method is quite good.

**Comment by martinrotter** [comment]
Pushed now, it should be compiled in 15 minutes.

**Comment by rfkat** [comment]
> Please test [daf0ca4](https://github.com/martinrotter/rssguard/commit/daf0ca46367c951964563db5b771adc1517b4a17) after it finishes building via development build and report back, thanks.

Tested using rssguard-devbuild-0716fd476-linux64.AppImage, Browsing the article filter list is as responsive as expected (instant response). Switching between labels takes around 8 seconds from clicking on the label until the articles show up.

**Comment by martinrotter** [comment]
> Changing from label to label

So what remains is problems when chaning among labels and when processing/testing filters in filters window. Pls report all slowness you see. I will first identify all problematic places and then address them somehow.

**Comment by martinrotter** [comment]
Working on this. It is big change but DB migration will be done on-the-fly and no major RSS Guard version change is needed.

**Comment by martinrotter** [comment]
OKAY. some code in the branch. I reworked database structure a bit. Now each articles has column "labels" which keeps list if activated labels per each article. This approach does speed up basically all operations related to labels. Majority of code is now untested which I will do next days. Note that database schema is now upgraded to version "5" so backup your database files prior testing.

https://github.com/martinrotter/rssguard/blob/solve-labels-performance/resources/sql/db_update_sqlite_4_5.sql

The DB schme upgrade is seamless and users will not even notice that happening. Report back any findings. Some label op migh be temporarily broken but will test all and fix.

**Comment by martinrotter** [comment]
RSS Guard runs filters against all articles downloaded from feed.

1. Feed file is downloaded.
2. Articles from feed file are extracted.
3. Articles are ran with filter if any is activated.
4. "accepted" articles are saved/overwritten in the DB.

As for "RSS Guard 4" folder. Yes, in RSS Guard 3 and before the folder was just "RSS Guard" but in RSS Guard 4 some major overhaul of database was done and it was not possible to automatically migrate old database files, therefore I decided to use completely new folder to save RSS Guard user data.

**Comment by martinrotter** [comment]
I tested selecting of articles with your "extreme" database and it is instant. Can you provide sample database (possibly whole "RSS Guard" folder which exibits the issue?

**Comment by martinrotter** [comment]
Also, I added a fix so sqlite database file backups are new versions like this: `database.db-v4.bak`

**Comment by martinrotter** [comment]
> Yes, I see now that filters only apply to incoming articles. I set up a bunch of filters, let them be enabled overnight, and saw that they hadn't applied labels to most things in my feeds. So I had to go and apply them manually. I should have tested this first.

Filters are applied to all incoming articles but if that particular article already exists in DB and is not "updated" then the "filtered" new "clone" is not stored in DB and original article resides there. If article is "updated" (that means that its title, creation date etc. changes), then it is overwritten in DB).

If you make some changes to filters which you need to apply to existing articles then you have to run "Process checked feeds" button in "Article filters" dialog.

**Comment by martinrotter** [comment]
Investigated, yes, the problem is linearly dependent on number of labels and problem is that when any UNREAD message is selected then counts of unread articles in each label might change, thus, they are recalculated, have to come with efficient strategy

**Comment by martinrotter** [comment]
4784baad4 - much faster right click on several articles -> label's context menu was the culprit

now, i will make it work much faster when browsing unread articles

**Comment by martinrotter** [comment]
The delay should be only with UNread articles (the marking as read is what takes the time now).

**Comment by aleksejrs** [comment]
4.3.4 (supposedly) often fails to mark the article read from reading it. I believe I mostly use left-click.

**Comment by martinrotter** [comment]
> 4.3.4 (supposedly; I am not sure I haven't built anything later) often fails to mark the article read from reading it. I believe I mostly use left-click.

Report as separate issue, but first test with latest development release.

**Comment by martinrotter** [comment]
> So it feels like it's just this final little bit and this thing could all be over.

Yes, that is the work I do now, should be ready soon.

**Comment by martinrotter** [comment]
OK, just commited.

d866378df

Please test when compiled and report back - now the performance when browsing article list (and marking as read in the proces) should be quite faster, but depending on number of labels assigned to affected articles.

We are now almost bump into limits of SQLite full-text search which is now used to detect assigned labels to each article. This could be now also HUGELY speeed up by using special FTS5 functions of SQLite databases, but thing is that those are not present in MariaDB which is supported too.

Anyway test and let me know if the performance is now satisfactory. It is fair to say that big majority of users will not have hundred of labels, but maybe 10-20 at most.

**Comment by aleksejrs** [comment]
> d866378

a category with 100k unread articles: ≈20 seconds.
an unread article: immediate!
a read article: ≈3 seconds.

**Comment by aleksejrs** [comment]
> an unread article: immediate! And the article is marked read.

They are not always marked read. A guess: maybe that happens when it takes a long time to open?

**Comment by martinrotter** [comment]
OK guys. I am closing this as solved. There are some more boosts which might be done, but MariaDB support would have to be dropped first which is not feasible at this point.

For 99 % of users, 1-100 labels with some thousands of articles is super fine and with these amounts it now works okay.

Report any problems in separated tickets.
